omegaconf/basecontainer.py

```diff
@@ -298,19 +299,41 @@ def expand(node: Container) -> None:
                 else:
                     node._set_value(type_)
 
-        if src._is_missing() and not dest._is_missing():
-            if is_structured_config(get_ref_type(src)):
-                expand(src)
+        src_ref_type = get_ref_type(src)
+        if (
+            src._is_missing()
+            and not dest._is_missing()
+            and is_structured_config(src_ref_type)
+        ):
+            assert src_ref_type is not None
+            # Replace `src` with a prototype of its corresponding structured config
+            # whose fields are all missing (to avoid overwriting fields in `dest`).
+            src_data = get_structured_config_data(src_ref_type)
+            for v in src_data.values():
+                v._set_value(MISSING)
+            src = DictConfig(src_data)
+            src._metadata.ref_type = src_ref_type
+            src._metadata.object_type = src_type
 
         if (dest._is_interpolation() or dest._is_missing()) and not src._is_missing():
             expand(dest)
 
         for key, src_value in src.items_ex(resolve=False):
             src_node = src._get_node(key, validate_access=False)
             dest_node = dest._get_node(key, validate_access=False)
-            if isinstance(dest_node, Container) and OmegaConf.is_none(dest, key):
-                if not OmegaConf.is_none(src_value):
-                    expand(dest_node)
+
+            if isinstance(src_value, Container):
+                missing_src_value = src_value._value() == MISSING
+            else:
+                missing_src_value = src_value == MISSING
```

Comment:
This too, a small helper will make this complex function a bit more readable.
```python
missing_src_value = missing_src_value(src_value)
```

Reply:
Done in 678a186